core/src/main/java/io/questdb/cairo/TxnScoreboardV2.java

```diff
@@ -0,0 +1,267 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Numbers;
+import io.questdb.std.Unsafe;
+import io.questdb.std.Vect;
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * In-memory transaction scoreboard. Each table reader mutates its own
+ * slot based on the assigned id. Cross-reader operations require all slots
+ * to be checked. On the other hand, single-reader operations are cheap.
+ */
+public class TxnScoreboardV2 implements TxnScoreboard {
+    private static final int RESERVED_ID_COUNT = 16;
+    private static final long UNLOCKED = -1;
+    private static final int VIRTUAL_ID_COUNT = 1;
+    private final int entryScanCount;
+    private final int pow2EntryCount;
+    private long activeReaderCountMem;
+    private long entriesMem;
+    private long maxMem;
+    private long maxReaderIdMem;
+    // Record structure
+    // 8 bytes - active reader count
+    // 8 bytes - max txn
+    // 8 bytes - max id
+    // 12 * 8 - reserved
+    // 8 bytes - slot for CHECKPOINT txn
+    // N * 8 bytes - slots for every TableReader
+    private long mem;
+
+    private TableToken tableToken;
+
+    public TxnScoreboardV2(int entryCount) {
+        pow2EntryCount = Numbers.ceilPow2(entryCount + RESERVED_ID_COUNT);
```

Comment:
Do we really need `ceilPow2` here considering that this is no longer a ring buffer-like data structure?

Reply:
No need, removed in [a3b5704](https://github.com/questdb/questdb/pull/5444/commits/a3b5704933449386a2f6a399ce5b4604838bf72e)